[CI][CUDA] Add periodic b200 distributed job

nWEIdia opened the pull request:
1. Run distributed job with B200 runner, periodically. 
2. discovered generic distributed test issue that certain unit test hard-coded ranks, calling for require_exact_world_size(world_size) API instead of require_world_size(world_size). 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @ptrblck @eqy @tinglvv @huydhn @atalman @ZainRizvi @malfet

nWEIdia committed:
Add b200 distributed job

Add aws-role-to-assume: arn:aws:iam::308535385114:role/gha_workflow_s3_and_ecr_read_only
to test job on B200 runner

Fix runner usage - this distributed job should use linux.dgx.b200.8

@require_world_size(4) does not translate to world_size==4.
For example, 8GPU B200 runner would make the unit test run with
world_size = 8.
Tested with:
TEMP_DIR=/tmp BACKEND=nccl WORLD_SIZE=4 pytest -v test/distributed/test_distributed_spawn.py -k test_new_subgroups_world_size_not_divisible_by_group_size

Add require_exact_world_size for distributed unit tests that implicitly
require world_size of 4 to pass.

Fix test_3_level_hierarchical_model_averager

Mimic H100 distributed, run distributed less often because it takes
quite long to finish all the tests (easily 4hours+ for each of the 3
shards).

huydhn commented on .github/workflows/pull.yml:
```diff
@@ -271,6 +271,23 @@ jobs:
         ]}
     secrets: inherit
 
+  linux-jammy-cuda12_8-py3_10-gcc11-build-distributed-b200:
```
This doesn't make sense to me yet, but we could chime in people from distributed team (cc  @kwen2501).  The problem here is that `linux.dgx.b200` has just 1 B200 GPU, so most distributed tests would be skipped anyway.  We could use `linux.dgx.b200.8`, but there are only 2 of them atm.  So, I don't think there are enough capacity to run it here.  We would need something like https://github.com/pytorch/pytorch/blob/main/.github/workflows/h100-distributed.yml that runs periodically.

nWEIdia replied:
good point, yes I missed that part. 
Indeed, distributed tests must have been mostly skipped.

nWEIdia replied:
As discussed, moving to periodic job to avoid the big queue.

nWEIdia commented:
Link https://github.com/pytorch/pytorch/pull/158695 for addressing the following failure also seen on B200: https://github.com/pytorch/pytorch/actions/runs/16757750565/job/47447831792 

 `AttributeError: '_MeshEnv' object has no attribute 'create_child_mesh'
2025-08-05T20:20:16.5442776Z 
2025-08-05T20:20:16.5442911Z To execute this test, run the following from the base repo dir:
2025-08-05T20:20:16.5443085Z     python test/distributed/checkpoint/e2e/test_fsdp_ep.py TestFSDPWithEP.test_e2e
2025-08-05T20:20:16.5443088Z 
2025-08-05T20:20:16.5443304Z This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0. Did you mean: 'create_sub_mesh'?`

nWEIdia commented:
python test/distributed/test_distributed_spawn.py TestDistBackendWithSpawn.test_3_level_hierarchical_model_averager failure in https://github.com/pytorch/pytorch/actions/runs/16764647752/job/47468897559 is because: 

The test originally expected len(period_group_size_dict) groups, but that's incorrect. The actual count depends on how many subgroups are created for each group size. For world_size=4: subgroup_size=2 creates 2 groups, subgroup_size=4 creates 0 new groups → total 2, total pg_group would be 3. For world_size=8: subgroup_size=2 creates 4 groups, subgroup_size=4 creates 2 groups → total 6, total pg_group would be 7

huydhn commented:
Looking at the number from https://github.com/pytorch/pytorch/actions/runs/16814299990/job/47629596189, I think we need to consider running this even less frequent, maybe nightly.  My logic is that we have 3 8xB200 runners here, so the total bank we have a day is 3 * 24 = 72 hours.  Each distributed shards from https://github.com/pytorch/pytorch/actions/runs/16814299990/job/47629596189 seem to take close to 4 hours (I'm not sure why they take that long to finish), there are 3 of them set to run every 4 hour, so 3 * 4 hours * 6 times will eat up all the capacity.  I'm proposing that we do the same as H100 distributed job to run just some smoke tests https://github.com/pytorch/pytorch/blob/main/.github/workflows/h100-distributed.yml#L46.

If we take a rough approach of dividing 72 into 4 groups for PyTorch core, compiler, distributed, and vLLM.  Distributed jobs shouldn't take more than 25% or 18 hours per day.

nWEIdia commented:
Yes, I think using 4XB200 may make a whole bunch of failed unit test pass automatically. 
Running nightly with just linux.dgx.b200.4 seems to be better than periodic with linux.dgx.b200.8.

nWEIdia committed:
Use linux.dgx.b200.8 since linux.dgx.b200.4 is not available.

nWEIdia committed:
Revert "Use linux.dgx.b200.8 since linux.dgx.b200.4 is not available."

This reverts commit 1ba7b2e20b804f337aa483e0228228273ade7396.

nWEIdia committed:
Check if using linux.dgx.b200.8 would avoid the fsdp issues.

nWEIdia commented:
@pytorchbot rebase

nWEIdia committed:
Undo changes to test_3_level_hierarchical_model_averager

nWEIdia committed:
Skip distributed unit tests with their issue numbers.

nWEIdia committed:
Fix Lint

nWEIdia committed:
Change runner to linux.12xlarge.memory
and fix lint

nWEIdia committed:
rebase

nWEIdia committed:
Further increase timeout.
test_gather_object* require exact world size (4) to succeed
Likely, all NVSHMEM Triton tests would fail. Skipping all for now.

nWEIdia committed:
Resolve merge conflicts

nWEIdia committed:
Resolve more merge conflicts

nWEIdia commented:
Oops, the distributed jobs seem to have hang after rebasing. 
https://github.com/pytorch/pytorch/actions/runs/17772045138/job/50513508855 this uses runner `dgxb200-05`,  I haven't updated the driver to 580.82.07 yet, so it is unclear why.

Aidyn-A commented:
I am debugging ` distributed/test_cupy_as_tensor.py::CupyAsTensorTest::test_cupy_as_tensor` failure. It is reproducable on non B200 machines as well (got the repro on A100x8). Seems like it is related to the expandable segments.

Aidyn-A committed:
skip test on SM100OrLater

Aidyn-A committed:
fix lint

Aidyn-A committed:
xfail test_fsdp_ep.py::TestFSDPWithEP::test_e2e

Aidyn-A committed:
skip test_cupy_as_tensor

Aidyn-A committed:
revert expectedFailure on TestFSDPWithEP::test_e2e

eqy commented:
@pytorchmergebot merge
